llvm/lib/Target/AMDGPU/GCNSubtarget.cpp

```diff
@@ -629,12 +623,16 @@ GCNUserSGPRUsageInfo::GCNUserSGPRUsageInfo(const Function &F,
       DispatchID = true;
   }
 
-  // TODO: This could be refined a lot. The attribute is a poor way of
-  // detecting calls or stack objects that may require it before argument
-  // lowering.
+  const bool IsNoFlatScratchInitSet =
+      F.hasFnAttribute("amdgpu-no-flat-scratch-init");
+
   if (ST.hasFlatAddressSpace() && AMDGPU::isEntryFunctionCC(CC) &&
       (IsAmdHsaOrMesa || ST.enableFlatScratch()) &&
-      (HasCalls || HasStackObjects || ST.enableFlatScratch()) &&
+      // The line below: If enableFlatScratch() is true, whether
+      // no-flat-scratch-init is set is not important. If enableFlatScratch()
+      // is false, FlatScratchInit cannot be true for graphics CC.
+      (ST.enableFlatScratch() ||
```

Comment:
```suggestion
      // FlatScratchInit cannot be true for graphics CC.
      (ST.enableFlatScratch() ||
```

This sounds broken. Flat addressing should work fine in shaders. The only difference will be in the sequence used to initialize flat_scr needs to be different. But that's a problem for later.

Also can sink the hasFnAttribute down here.

Reply:
I mean inline in the expression. Also "cannot be true", rephrase are something about the graphics shaders not having the input

Reply:
> I mean inline in the expression (and swap the order, CC check first). Also "cannot be true", rephrase are something about the graphics shaders not having the input

What does "not having the input" mean?